File: app/src/main/java/com/larryhsiao/nyx/jots/JotsByLocation.java
package com.larryhsiao.nyx.jots;

import com.silverhetch.clotho.Source;
import org.locationtech.jts.geom.Geometry;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

/**
 * Source to find jots by given geometry polygon.
 */
public class JotsByLocation implements Source<ResultSet> {
    private final Source<Connection> conn;
    private final Geometry geometry;

    public JotsByLocation(Source<Connection> conn, Geometry geometry) {
        this.conn = conn;
        this.geometry = geometry;
    }

    @Override
    public ResultSet value() {
        try {
            PreparedStatement stmt = conn.value().prepareStatement(
                // language=H2
                "SELECT * FROM jots " +
                    "WHERE location && ?" +
                    "ORDER BY CREATEDTIME DESC;"
            );
            stmt.setString(1, geometry.toText());
            return stmt.executeQuery();
        } catch (SQLException e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: core/src/main/java/com/larryhsiao/nyx/core/jots/JotsByLocation.kt
package com.larryhsiao.nyx.core.jots

import com.silverhetch.clotho.Source
import org.locationtech.jts.geom.Geometry
import java.sql.Connection
import java.sql.ResultSet
import java.sql.SQLException

/**
 * Source to find jots by given geometry polygon.
 */
class JotsByLocation(
    private val conn: Source<Connection>,
    private val geometry: Geometry
) : Source<ResultSet> {
    override fun value(): ResultSet {
        return try {
            val stmt = conn.value().prepareStatement( // language=H2
                "SELECT * FROM jots " +
                    "WHERE location && ? " +
                    "AND DELETE = 0 " +
                    "ORDER BY CREATEDTIME DESC;"
            )
            stmt.setString(1, geometry.toText())
            stmt.executeQuery()
        } catch (e: SQLException) {
            throw IllegalArgumentException(e)
        }
    }
}